src/tls_wrap.cc

```diff
@@ -881,6 +889,90 @@ int TLSWrap::SelectSNIContextCallback(SSL* s, int* ad, void* arg) {
 #endif  // SSL_CTRL_SET_TLSEXT_SERVERNAME_CB
 
 
+#ifdef OPENSSL_PSK_SUPPORT
+
+unsigned int TLSWrap::PskServerCallback(SSL *ssl,
+                                        const char *identity,
+                                        unsigned char *psk,
+                                        unsigned int max_psk_len) {
+  TLSWrap* p = static_cast<TLSWrap*>(SSL_get_app_data(ssl));
+  Environment* env = p->env();
+
+  Local<Value> argv[] = {
+    String::NewFromUtf8(env->isolate(), identity)
+  };
+  Local<Value> ret = p->MakeCallback(env->onpskexchange_string(),
+                                     arraysize(argv),
+                                     argv);
+
+  // The result is expected to be an object with a "psk" buffer value. If this
+  // isn't the case then return 0, indicating that the identity isn't found.
+  if (!ret->IsObject()) {
+    return 0;
+  }
+  Local<Object> obj = ret.As<Object>();
+
+  Local<Value> psk_buf = obj->Get(env->psk_string());
+  if (!Buffer::HasInstance(psk_buf)) {
+    return 0;
+  }
+
+  size_t psk_len = Buffer::Length(psk_buf);
+  if (psk_len > max_psk_len) {
+    // FIXME: emit an error
+    return 0;
+  }
+
+  memcpy(psk, Buffer::Data(psk_buf), psk_len);
+  return psk_len;
+}
+
+unsigned int TLSWrap::PskClientCallback(SSL *ssl,
+                                        const char *hint,
+                                        char *identity,
+                                        unsigned int max_identity_len,
+                                        unsigned char *psk,
+                                        unsigned int max_psk_len) {
+  TLSWrap* p = static_cast<TLSWrap*>(SSL_get_app_data(ssl));
+  Environment* env = p->env();
+
+  Local<Value> argv[] = { Null(env->isolate()) };
+  if (hint != nullptr)
+      argv[0] = String::NewFromUtf8(env->isolate(), hint);
+  Local<Value> ret = p->MakeCallback(env->onpskexchange_string(),
+                                     arraysize(argv),
+                                     argv);
+
+  // The result is expected to be an object with "identity" string and "psk"
+  // buffer values. If this isn't the case then return 0, indicating that the
+  // identity isn't found.
+  if (!ret->IsObject()) {
+    return 0;
+  }
+  Local<Object> obj = ret.As<Object>();
+
+  Local<Value> identity_str = obj->Get(env->identity_string());
+  Local<Value> psk_buf = obj->Get(env->psk_string());
+  if (!identity_str->IsString() || !Buffer::HasInstance(psk_buf)) {
+    return 0;
```

Comment:
This callback returns a psk length so that returning 0 leads TLS fatal alert of unknown_psk_identity. I think throwing error is better for invalid inputs.